### PR TITLE
Add code reloading to daemon loop for debug execution.

### PR DESCRIPTION
**Pull request by jessemyers**
The Flask development server detects code changes and reloads; now daemon event loops can do the same.

### CODE REVIEW

File: microcosm_daemon/reloader.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+Cribbed in part from: https://github.com/pallets/werkzeug/blob/master/werkzeug/_reloader.py
+
+"""
+from logging import getLogger
+import os
+import sys
+
+
+def _iter_module_files():
```
Comment (jessemyers) [inline microcosm_daemon/reloader.py]: This is a verbatim copy from werkzeug.

File: microcosm_daemon/reloader.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+class Reloader(object):
+    """
+    Checks modification times for Python modules and reloads execution.
+
+    """
+    def __init__(self):
+        self.mtimes = {}
+
+    @property
+    def changed(self):
```
Comment (jessemyers) [inline microcosm_daemon/reloader.py]: This is very similar to the equivalent code in `werkzeug` except that I wrote it as a `@property` instead of a loop since the daemon infrastructure already has its own loop.

File: microcosm_daemon/reloader.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    def reexecute(self):
+        """
+        Re-execute the current program.
+
+        """
+        getLogger().info("Reloading executable: '{} {}'".format(
+            sys.executable,
+            " ".join(sys.argv),
+        ))
+        os.execve(sys.executable, [sys.executable] + sys.argv, os.environ)
```
Comment (jessemyers) [inline microcosm_daemon/reloader.py]: To the best of my knowledge, this will re-execute the current program. It definitely works in my tests, but there could be some esoteric edge cases.

File: microcosm_daemon/state_machine.py [hunk 4]
```diff
@@ -50,5 +52,7 @@ def run(self):
             with self.graph.signal_handler:
                 while self.should_run():
                     self.current_state = self.step()
+                    if self.reloader:
```
Comment (jessemyers) [inline microcosm_daemon/state_machine.py]: The event loop checks the reloader on every loop iteration (but only if `debug` is enabled).

File: microcosm_daemon/state_machine.py [hunk 5]
```diff
@@ -10,9 +11,10 @@ class StateMachine(object):
 
     """
 
-    def __init__(self, graph, initial_state):
+    def __init__(self, graph, initial_state, never_reload=False):
```
Comment (jessemyers) [inline microcosm_daemon/state_machine.py]: I though it'd be nice to support disabling reloading, but perhaps this is YAGNI?

**Comment by adamhadani** [comment]
LGTM
